### PR TITLE
Allows RedisClusterProxy instance in Lock RedisStore

### LINKED ISSUE #37476: RedisStore accepts \RedisCluster but not RedisClusterProxy
**Symfony version(s) affected**: tested with 5.1

**Description**  

Symfony\Component\Lock\Store\RedisStore::__construct() accepts instance of `\RedisCluster` as argument but it does not accept `Symfony\Component\Cache\Traits\RedisClusterProxy`. Since `Symfony\Component\Lock\Store\StoreFactory` forces lazy connections when creating Redis client, you will always get RedisClusterProxy instead of \RedisCluster if your DSN has `?redis_cluster=1`.

We are using Redis cluster and when we switched from Predis to php-redis, it resulted in following error because of this:
_Error: ""Symfony\Component\Lock\Store\RedisStore::__construct()" expects parameter 1 to be Redis, RedisArray, RedisCluster or Predis\ClientInterface, "Symfony\Component\Cache\Traits\RedisClusterProxy" given."_

I understand Redis cluster is not reliable for storing locks but if RedisStore accepts \RedisCluster as argument, it should also accept decorator RedisClusterProxy.

**How to reproduce**  

Using Redis as lock store, add `?redis_cluster=1` to your lock store DSN.

**Possible Solution**  

Update RedisStore::__construct() to accept RedisClusterProxy as argument.

Another option if Redis cluster should never be used with Lock, remove \RedisCluster and RedisClusterProxy references from RedisStore and StoreFactory::createStore() to avoid confusion.

**Additional context**

### PR DESCRIPTION
**Pull request by jderusse**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #37476
| License       | MIT
| Doc PR        | /

This PR allow clients instance of RedisClusterProxy created by the StoreFactory in #28691 /cc @nicolas-grekas

### CODE REVIEW

File: src/Symfony/Component/Cache/Traits/RedisTrait.php [hunk 1]
```diff
@@ -84,7 +84,7 @@ private function init($redisClient, string $namespace, int $defaultLifetime, ?Ma
      *
      * @throws InvalidArgumentException when the DSN is invalid
      *
-     * @return \Redis|\RedisCluster|\Predis\ClientInterface According to the "class" option
+     * @return \Redis|\RedisCluster|RedisClusterProxy|RedisProxy|\Predis\ClientInterface According to the "class" option
```
Comment (jderusse) [inline src/Symfony/Component/Cache/Traits/RedisTrait.php:87]: I'm not sure if this doc block should be fixed

**Comment by jderusse** [comment]
failing Travis test is related to branch 3.4 (when Redis server is not running)
failing github test is related to  cache

**Review by nicolas-grekas — APPROVED** [review]

File: src/Symfony/Component/Cache/Traits/RedisTrait.php (hunk 1 above)
Comment (stof) [reply src/Symfony/Component/Cache/Traits/RedisTrait.php:87]: well, if we don't update it, the implementation would not respect the contract of the signature. So updating it is right.

File: src/Symfony/Component/Lock/Tests/Store/StoreFactoryTest.php [hunk 2]
```diff
@@ -61,6 +61,8 @@ public function validConnections()
         }
         if (class_exists(\Redis::class) && class_exists(AbstractAdapter::class)) {
             yield ['redis://localhost', RedisStore::class];
+            yield ['redis://localhost?redis_cluster=1', RedisStore::class];
+            yield ['redis://localhost?lazy=1', RedisStore::class];
```
Comment (stof) [inline src/Symfony/Component/Lock/Tests/Store/StoreFactoryTest.php:65]: what about the lazy cluster ?

**Comment by nicolas-grekas** [comment]
Are the failures related? Can you rebase to retrigger maybe?

**Comment by jderusse** [comment]
> Are the failures related? Can you rebase to retrigger maybe?

rebased. Faillure are not related.

failure is about test against branch 3.4 without Redis container running nor mechanism to early exit when redis is not runing

**Comment by jderusse** [comment]
CI green @nicolas-grekas

**Comment by nicolas-grekas** [comment]
Thank you @jderusse.

File: src/Symfony/Component/Lock/Store/RedisStore.php [hunk 3]
```diff
@@ -33,13 +33,13 @@ class RedisStore implements StoreInterface
     private $initialTtl;
 
     /**
-     * @param \Redis|\RedisArray|\RedisCluster|\Predis\ClientInterface $redisClient
-     * @param float                                                    $initialTtl  the expiration delay of locks in seconds
+     * @param \Redis|\RedisArray|\RedisCluster|RedisProxy|RedisClusterProxy\Predis\ClientInterface $redisClient
```
Comment (b1rdex) [inline src/Symfony/Component/Lock/Store/RedisStore.php:36]: @jderusse There is a bug: no pipe in `RedisClusterProxy\Predis\ClientInterface` leads to wrong class FQN.

**Pull request merged**